recipes-app/iot2050-firmware-update/files/iot2050-firmware-update

```diff
@@ -118,61 +178,65 @@ class FirmwareUpdate(object):
-
-                if not mtd_content == firmware_content:
-                    print("U", end="")
-                    sys.stdout.flush()
-                    self.flash_erase(mtd_dev, mtd_pos, mtd_erasesize)
-                    os.lseek(mtd_dev, mtd_pos, os.SEEK_SET)
-                    os.write(mtd_dev, firmware_content)
-                else:
-                    print(".", end="")
-                    sys.stdout.flush()
-                mtd_pos += mtd_erasesize
-                firmware_size -= mtd_erasesize
-            print()
-            os.close(mtd_dev)
-
-            mtd_num += 1
+        self.firmware_tar.seek(0)
+        with tarfile.open(fileobj=self.firmware_tar).extractfile(self.firmware_name) as firmware:
+            firmware.seek(0)
+            firmware.seek(0, os.SEEK_END)
+            firmware_size = firmware.tell() 
+            firmware.seek(0)
+
+            while True:
+                if firmware_size <= 0:
+                    print("\nCompleted. Please reboot the device\n")
+                    sys.exit(0)
+
+                mtd_sys_path = "{}/mtd{}".format(mtd_base_path, mtd_num)
+                mtd_name_path = "{}/name".format(mtd_sys_path)
+                mtd_size_path = "{}/size".format(mtd_sys_path)
+                mtd_erasesize_path = "{}/erasesize".format(mtd_sys_path)
+                mtd_dev_path = "/dev/mtd{}".format(mtd_num)
+
+                mtd_size = int(self.get_path_type_value(mtd_size_path))
+                mtd_erasesize = int(self.get_path_type_value(mtd_erasesize_path))
+                mtd_name = self.get_path_type_value(mtd_name_path).strip()
+                mtd_pos = 0
+
+                print("Updating %-20s" % mtd_name, end="")
+
+                try:
+                    mtd_dev = os.open(mtd_dev_path, os.O_SYNC | os.O_RDWR)
+                except IOError as e:
+                    print("Opening {} failed: {}".format(mtd_dev_path, e.strerror))
+                    sys.exit(1)
+
+                while mtd_pos < mtd_size and firmware_size > 0:
+                    mtd_content = os.read(mtd_dev, mtd_erasesize)
+                    firmware_content = firmware.read(mtd_erasesize)
+                    padsize = mtd_erasesize - len(firmware_content)
+                    firmware_content += bytearray([0xff] * padsize)
+
+                    if not mtd_content == firmware_content:
+                        print("U", end="")
+                        sys.stdout.flush()
+                        self.flash_erase(mtd_dev, mtd_pos, mtd_erasesize)
+                        os.lseek(mtd_dev, mtd_pos, os.SEEK_SET)
+                        os.write(mtd_dev, firmware_content)
+                    else:
+                        print(".", end="")
+                        sys.stdout.flush()
+                    mtd_pos += mtd_erasesize
+                    firmware_size -= mtd_erasesize
+                print()
+                os.close(mtd_dev)
+
+                mtd_num += 1
```

Comment:
This is the heart of the update tool. Can you refactor this separately for the new feature?

Reply:
Sorry, not fully follow you, do you suggest to separate the new tarball feature out of the `FirmwareUpdate` class?

Reply:
I only request to factor our the change to this many function into a separate commit, explaining what it changes (replace variable name? or more?). It's hard to review.

Reply:
I think the whitespace change make it hard to review, please toggle the `ignore white space` from the gear icon